contracts/eosio.system/src/finalizer_key.cpp

```diff
@@ -115,42 +125,45 @@ namespace eosiosystem {
 
       // Basic key and signature format check
       check(finalizer_key.compare(0, 7, "PUB_BLS") == 0, "finalizer key must start with PUB_BLS");
-      check(proof_of_possession.compare(0, 7, "SIG_BLS") == 0, "proof of possession siganture must start with SIG_BLS");
-
-      // Convert to binary form
-      const auto fin_key_g1 = eosio::decode_bls_public_key_to_g1(finalizer_key);
-      const auto signature = eosio::decode_bls_signature_to_g2(proof_of_possession);
+      check(proof_of_possession.compare(0, 7, "SIG_BLS") == 0, "proof of possession signature must start with SIG_BLS");
 
       // Duplication check across all registered keys
       auto idx = _finalizer_keys.get_index<"byfinkey"_n>();
       auto hash = get_finalizer_key_hash(finalizer_key);
       check(idx.find(hash) == idx.end(), "duplicate finalizer key");
 
-      // Proof of possession check is expensive, do it at last
-      check(eosio::bls_pop_verify(fin_key_g1, signature), "proof of possession check failed");
-
-      // Insert into finalyzer_keys table
-      auto key_itr = _finalizer_keys.emplace( finalizer_name, [&]( auto& k ) {
-         k.id                 = _finalizer_keys.available_primary_key();
-         k.finalizer_name     = finalizer_name;
-         k.finalizer_key      = finalizer_key;
-         k.finalizer_key_hash = get_finalizer_key_hash(finalizer_key);
+      // Convert to binary form. The validity will be checked during conversion.
+      const auto fin_key_g1 = eosio::decode_bls_public_key_to_g1(finalizer_key);
+      const auto pop_g2 = eosio::decode_bls_signature_to_g2(proof_of_possession);
+
+      // Proof of possession check is expensive, do it last
```

Comment:
What is the rational for doing to last because it is expensive. I think you could argue you should do it first because it is expensive.